File: String/Easy/20. Valid Parentheses.go
package main

import (
	"fmt"
	"strings"
)

func main() {

	str := "{[(]}"
	result := isValid(str)
	fmt.Println(result)
}

func isValid(s string) bool {
	str, temp := strings.Split(s, ""), ""
	var stack []string

	for i := 0; i < len(str); i++ {
		if str[i] == "(" || str[i] == "[" || str[i] == "{" {
			stack = append(stack, str[i])
			continue
		}
		if len(stack) == 0 {
			return false
		}
		switch str[i] {
		case ")":
			temp = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if temp != "(" {
				return false
			}
			break
		case "]":
			temp = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if temp != "[" {
				return false
			}
			break
		case "}":
			temp = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if temp != "{" {
				return false
			}
			break
		}
	}
	if len(stack) == 0 {
		return true
	}
	return false
}

File: Miscellaneous/Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, str):
        stack = []

        for i in range(len(str)):
            if str[i] == "(" or str[i] == "[" or str[i] == "{":
                stack.append(str[i])
                continue
            else:
                temp = stack.pop()
                if str[i] == ")" and (temp != "("):
                    return False
                elif str[i] == "]" and (temp!= "["):
                    return False
                elif str[i] == "}" and (temp != "{"):
                    return False
        
        if len(stack) == 0:
            return True
        return False

    def isValid2(self, str):
        stack = []
        dict = {")":"(", "}":"{","]":"["}

        for char in str:
            if char in dict.values():
                stack.append(char)
                continue
            elif char in dict.keys():
                if stack == [] or dict[char] != stack.pop():
                    return False
            else:
                return False

        return stack == []

obj = Solution()
ans = obj.isValid2("(([]{}))")
print(ans)
                    